theano/tensor/signal/downsample.py

```diff
@@ -518,6 +600,15 @@ def __init__(self, ds, ignore_border, st=None, padding=(0, 0), mode='max'):
                 " 'average_inc_pad' and 'average_exc_pad'. Got %s" % mode)
         self.mode = mode
 
+    def infer_shape(self, node, in_shapes):
+        return [in_shapes[0]]
+
+
+class MaxPoolGrad(PoolGrad):
```

Comment:
I don't really follow and I never did an optimization before. Some extra details would be helpful.